File: app/day03/parts.js
const { readFile, strToInt } = require("../support/utils");

const REGEX = new RegExp(/(mul\((\d+),(\d+)\)|do\(\)|don't\(\))/g);

const compute_part_1 = () =>
  readFile("day03/input.txt")
    .matchAll(REGEX)
    .reduce((acc, item) => {
      const [_, op, n1, n2] = item;
      return op.includes("mul") ? acc + n1 * n2 : acc;
    }, 0);

const compute_part_2 = () => {
  let execute = true;
  let instructionsSum = 0;
  const items = readFile("day03/input.txt").matchAll(REGEX);
  items.forEach((item, idx) => {
    const [_, op, n1, n2] = item;
    if (op.includes("mul")) {
      instructionsSum += execute || idx == 0 ? strToInt(n1) * strToInt(n2) : 0;
    } else if (op.includes("don't")) {
      execute = false;
    } else {
      execute = true;
    }
  });
  return instructionsSum;
};

console.assert(compute_part_1() == 159833790);
console.assert(compute_part_2() == 89349241);

File: app/day03/parts.py
import re
import sys
from enum import Enum

from app.support.utils import main, parse_module_to_day, timing

REGEX = r = re.compile(r"(mul\((\d+),(\d+)\)|do\(\)|don't\(\))")


class Operation(Enum):
    MUL = "mul"
    DO = "do"
    DONT = "don't"


@timing
def compute_part_1(input_: str) -> int:
    return sum(int(n1) * int(n2) for op, n1, n2 in REGEX.findall(input_) if "mul" in op)


@timing
def compute_part_2(input_: str) -> int:
    execute = True
    instructions_sum = 0
    for idx, (op, n1, n2) in enumerate(REGEX.findall(input_)):
        match Operation(op.split("(")[0]):
            case Operation.MUL:
                if execute or idx == 0:
                    instructions_sum += int(n1) * int(n2)
            case Operation.DO:
                execute = True
            case Operation.DONT:
                execute = False
    return instructions_sum


if __name__ == "__main__":  # pragma: no cover
    module, day = parse_module_to_day(sys.modules[__name__].__package__)

    main(day, module, compute_part_1, compute_part_2)
